File: src/functionalModel/TestModel.cpp
// #include <pybind11/pybind11.h>

#include "TestModel.hpp"

#define TestModel_L_dimension 4
#define TestModel_D_dimension 9

TestModel::TestModel()
{
    A_ = mat(TestModel_D_dimension, TestModel_L_dimension, fill::zeros); //! Seg fault on this line when calling ctest (checked with address sanitizer)
    A_ = {{1, 2, 2, 1},
               {0, 0.5, 0, 0},
               {0, 0, 1, 0},
               {0, 0, 0, 3},
               {0.2, 0, 0, 0},
               {0, -0.5, 0, 0},
               {-0.2, 0, -1, 0},
               {-1, 0, 2, 0},
               {0, 0, 0, -0.7}};
    A_ *= 0.5;
}

void TestModel::F(vec x, vec &y)
{
    vec Hx(TestModel_L_dimension, fill::ones);
    vec Gx(TestModel_L_dimension, fill::ones);

    // Fill Hx
    Hx(0) = x(0);
    Hx(1) = x(1);
    Hx(2) = 4. * pow(x(2) - 0.5, 2);
    Hx(3) = x(3);

    // Fill Gx
    Gx(0) = exp(Hx(0));
    Gx(1) = exp(Hx(1));
    Gx(2) = exp(Hx(2));
    Gx(3) = exp(Hx(3));

    // compute y
    y = A_ * Gx;
}

unsigned TestModel::getDimensionY()
{
    return TestModel_D_dimension;
}

unsigned TestModel::getDimensionX()
{
    return TestModel_L_dimension;
}

void TestModel::toPhysic(vec &x)
{
    x *= 2;
}

void TestModel::fromPhysic(vec &x)
{
    x /= 2;
}

File: tests/dataRef/externalPythonModels/testmodel.py
import math

import numpy as np


class TestModel(object):
    def F(self, x):
        A = np.array([[1, 2, 2, 1],
                      [0, 0.5, 0, 0],
                      [0, 0, 1, 0],
                      [0, 0, 0, 3],
                      [0.2, 0, 0, 0],
                      [0, -0.5, 0, 0],
                      [-0.2, 0, -1, 0],
                      [-1, 0, 2, 0],
                      [0, 0, 0, -0.7]])
        A *= 0.5
        Hx = np.ones(self.get_L_dimension())
        Gx = np.ones(self.get_L_dimension())
        Hx[0] = x[0]
        Hx[1] = x[1]
        Hx[2] = 4. * math.pow(x[2] - 0.5, 2)
        Hx[3] = x[3]

        Gx[0] = math.exp(Hx[0])
        Gx[1] = math.exp(Hx[1])
        Gx[2] = math.exp(Hx[2])
        Gx[3] = math.exp(Hx[3])

        y = np.dot(A, Gx.transpose())
        return y

    def get_D_dimension(self):
        return 9

    def get_L_dimension(self):
        return 4

    def to_physic(self, x):
        return x

    def from_physic(self, x):
        return x

